### ISSUE TITLE
SequentialLR cannot be used with ReduceLROnPlateau due to .step() not allowing for optional arguments

### ISSUE THREAD

**Issue by Nesqulck** [pytorch/pytorch#68978]
## 🐛 Bug

Currently, SequentialLR can only be used with schedulers that inherit from _LRScheduler or in other words adhere to the way .step() is called, i.e. without any arguments. This is not the case for the built in ReduceLROnPlateau as it takes in a metric value. Therefore, when calling SequentialLR.step() without arguments, ReduceLROnPlateau will raise an error once its milestone is reached. However, when calling SequentialLR.step(metric), SequentialLR will raise an error due to too many arguments.

Also see https://github.com/PyTorchLightning/pytorch-lightning/issues/10759

## To Reproduce
```python
scheduler1 = ConstantLR(self.opt, factor=0.1, total_iters=2)
scheduler2 = ReduceLROnPlateau (self.opt, gamma=0.9)
scheduler = SequentialLR(self.opt, schedulers=[scheduler1, scheduler2], milestones=[2])
for epoch in range(100):
   train(...)
   validate(...)
   scheduler.step()  # raises error after epoch>=2
   scheduler.step(metric)  # raises error immediately
```

## Expected behavior
SequentialLR should allow for arbitrary arguments in step to be used with the underlying scheduler. Either the user takes care of passing the right argument at the right time or there is a built in mechanic (if that is even possible).

**Comment by jaketae** [comment]
I would like to take a jab at this if no one hasn't already!

Before I open a PR, I'd like to discuss a few proposals.

1. Convenient edge-casing

We can simply edge case `ReduceLROnPlateau` in `SequentialLR.step()` function. The step function should be able to take an optional argument `metrics` and will pass that argument to the scheduler if it is an instance of `ReduceLROnPlateau`. 

- Pros: Easy to implement
- Cons: Users might inherit from `ReduceLROnPlateau` in ways that may not be accounted for.

```python
def step(self, metrics=None):
        self.last_epoch += 1
        idx = bisect_right(self._milestones, self.last_epoch)
        scheduler = self._schedulers[idx]
        is_reduce_lr_on_plateau = isinstance(scheduler, ReduceLROnPlateau)
        is_new_scheduler = idx > 0 and self._milestones[idx - 1] == self.last_epoch
        if is_reduce_lr_on_plateau and is_new_scheduler:
            scheduler.step(metrics, 0)
        elif is_reduce_lr_on_plateau:
            scheduler.step(metrics)
        elif is_new_scheduler:
            scheduler.step(0)
        else:
            scheduler.step()
```

2. More general edge-casing

Make the step function even more general so that the schedulers do what they want to do with optional `args` and `kwargs`. Instead of checking if a scheduler is an instance of `ReduceLROnPlateau`, check if it is an instance of the `_LRScheduler` base class.

```python
def step(self, *args, **kwargs):
        self.last_epoch += 1
        idx = bisect_right(self._milestones, self.last_epoch)
        scheduler = self._schedulers[idx]
        is_lr_scheduler = isinstance(scheduler, _LRScheduler)
        is_new_scheduler = idx > 0 and self._milestones[idx - 1] == self.last_epoch
        if is_lr_scheduler and is_new_scheduler:
            scheduler.step(0)
        elif is_lr_scheduler:
            scheduler.step()
        elif is_new_scheduler:
            scheduler.step(*args, **kwargs, epoch=0)
        else:
            scheduler.step(*args, **kwargs)
```

3. Inheritance and modifying `_LRScheduler`

Given issues like https://github.com/pytorch/pytorch/issues/67760 and https://github.com/pytorch/pytorch/issues/68979, perhaps we should rewrite `ReduceLROnPlateau` to inherit from `_LRScheduler` and modify parts of the base class API. This direction would likely be the most time-consuming but arguably beneficial.

What do you think? @jbschlosser @albanD

**Comment by francois-rozet** [comment]
An alternative is to make the `step` function of `_LRScheduler` absorb and ignore all arguments (with `*args, **kwargs` in the signature) which allows to propagate the arguments of  `SequentialLR.step` without worry. The children of `_LRScheduler` would be responsible for using or not these arguments. There are several ways to do that.

For instance, we could add a method (say `get_get_lr`) in `_LRScheduler` that would absorb all arguments and return `self.get_lr()`. `step` would call this function instead of `get_lr`.

```python
def get_get_lr(self, *args, **kwargs) -> List[float]:
    return self.get_lr()
```

A scheduler with arguments would simply need to overwrite `get_get_lr`. For instance, this would allow to rewrite the `ReduceLROnPlateau` class as

```python
def lrs(optimizer: Optimizer) -> List[float]:
    return [group['lr'] for group in optimizer.param_groups]

class ReduceLROnPlateau(_LRScheduler):
    r"""Reduce learning rate when a metric has stopped improving"""

    def __init__(
        self,
        optimizer: Optimizer,
        gamma: float = 0.5,  # <= 1
        patience: int = 7,
        cooldown: int = 1,
        threshold: float = 1e-2,
        mode: str = 'minimize',  # or 'maximize'
        min_lr: Union[float, List[float]] = 1e-6,
        last_epoch: int = -1,
        verbose: bool = False,
    ):
        self.gamma = gamma
        self.patience = patience
        self.cooldown = cooldown
        self.threshold = threshold
        self.mode = mode

        if type(min_lr) is float:
            min_lr = [min_lr] * len(optimizer.param_groups)
        self.min_lrs = min_lr

        self.best = float('-inf') if self.mode == 'maximize' else float('inf')
        self.last_best = last_epoch
        self.last_reduce = last_epoch

        super().__init__(optimizer, last_epoch, verbose)

    def get_get_lr(self, last: float, *args, **kwargs) -> List[float]:
        return self.get_lr(last)

    def get_lr(self, last: float) -> List[float]:
        if self.mode == 'maximize':
            accept = last >= self.best * (1 + self.threshold)
        else:  # mode == 'minimize'
            accept = last <= self.best * (1 - self.threshold)

        if accept:
            self.best = last
            self.last_best = self.last_epoch

            return lrs(self.optimizer)

        if self.last_epoch - max(self.last_best, self.last_reduce + self.cooldown) <= self.patience:
            return lrs(self.optimizer)

        self.last_reduce = self.last_epoch

        return [
            max(lr * self.gamma, min_lr)
            for lr, min_lr in zip(lrs(self.optimizer), self.min_lrs)
        ]
```

What do you think? @albanD, @jaketae

**Comment by jaketae** [comment]
Hey @francois-rozet, thanks for the reply and suggestion! It definitely makes sense. I'm personally in support of any solution that prioritizes implementation simplicity and BC.

Gently pinging @albanD to see if there are any updates from the internal PyTorch team on this!

**Comment by imarquart** [comment]
Any updates on this?

**Comment by albanD** [comment]
No update I'm afraid.
The Core team is pretty small and we don't have anyone working on LRScheduler.

My general stance of LRScheduler is that they would require a much bigger (most likely BC-breaking) cleanup to get them in a position where they are reliable.
I would be happy to review a fix for this if they are small and BC but I'm not sure a complex change is worth it given the general state.

**Comment by AngledLuffa** [comment]
Just to ping about this without opening a new issue - is there any possibility of a PR which allows for a scheduler such as `ReduceLROnPlateau` to be used in a `Sequential...` or `ChainedScheduler`?

**Comment by abbas695** [comment]
is there anyone that found a workaround for this as the above solutions could not help me it says that the metrics is not passed and always none . but when using reducelronplateu alone it works

**Comment by MatZar01** [comment]
While it is not very elegant, a quick solution is to override step function in SequentialLR class to include metrics for ROP scheduler like this:

```
class SequentialLR_ROP(SequentialLR):
    def __init__(self, optimizer, schedulers, milestones):
        super().__init__(optimizer, schedulers, milestones)

    def step(self, metrics: Optional[int] = ..., epoch: Optional[int] = ...) -> None:
        self.last_epoch += 1
        idx = bisect_right(self._milestones, self.last_epoch - 1)  # -1 to ensure previous scheduler stopped when it should
        scheduler = self._schedulers[idx]
        if idx > 0 and self._milestones[idx - 1] == self.last_epoch:
            if scheduler.__class__.__name__ == 'ReduceLROnPlateau':
                scheduler.step(metrics)
            else:
                scheduler.step(0)
        else:
            if scheduler.__class__.__name__ == 'ReduceLROnPlateau':
                scheduler.step(metrics)
            else:
                scheduler.step()

        self._last_lr = scheduler.get_last_lr()
```

And then use `SequentialLR_ROP` in the same way as regular `SequentialLR`. I tested it with `schedulers=[LinearLR, ROP]` and it seems to be working fine.

**Comment by annalena-k** [comment]
Unfortunately, the quick solution by @MatZar01 does not work due to #121633.
Are there any updates on this issue?

**Comment by rmeit** [comment]
Has there been any progress on this issue? I've been getting the same series of errors using SequentialLR to chain LambdaLR and CosineAnnealingWarmRestarts schedulers.

**Comment by pupeno** [comment]
I ran into this at work and have opened two PRs to address it.

The first is a minimal implementation:

https://github.com/pytorch/pytorch/pull/195633

It follows the narrow `ReduceLROnPlateau` special-case direction discussed above, with the aim of changing as little as possible and keeping the necessary lifecycle handling consolidated in `ReduceLROnPlateau`. I have self-reviewed this branch and it is ready for review if this is a direction the project wants to take.

The second is a larger API proposal:

https://github.com/pytorch/pytorch/pull/195634

Rather than retrofit the legacy `ReduceLROnPlateau`, it introduces a composable `PlateauLR` that follows the regular `LRScheduler` lifecycle and takes the monitored value through a keyword-only `metrics` argument. The composing schedulers forward that argument to children that accept it, while legacy `ReduceLROnPlateau` keeps its existing API and is deprecated in favor of the new scheduler. This branch needs a more thorough line-by-line self-review, but if this direction seems likely to be accepted, I am happy to do that work promptly.

This is my first contribution to PyTorch. I would appreciate any feedback on the design or on how I can improve the contribution, and I hope it is the first of many.
